dom/src/main/scala/fs2/dom/Window.scala

```diff
@@ -67,7 +70,25 @@ object Window {
 
       implicit def given_Dom_F: Dom[F] = Dom.forAsync
 
-      def document: HtmlDocument[F] = window.document.asInstanceOf[HtmlDocument[F]]
+      def document: HtmlDocument[F] =
+        window.document.asInstanceOf[HtmlDocument[F]]
+
+      def domContentLoaded: F[Unit] = F.async_[Unit] { cb =>
+        // This _must_ happen in the same thunk because the event might fire before we can register an
+        // event handler
```

Comment:
Hmm originally I had something like the snippet,

```scala
for {
  state <- F.delay { dom.readyState }
  _ <- if (state == "loading") F.async_ { dom.addEventListener(...) } else IO.pure(())
} yield ()
```

in an `IOApp.Simple`, which in develop mode didn't seem to be triggering every so often in Firefox but working in Chrome. I _thought_ that this might be the issue, through the judicious use of `println` and since it did "fix" the behaviour I haven't investigated further.

If the cats effect runtime can guarantee that it won't happen, then I don't mind switching back to above implementation; but it really is quite unpleasant if it allows it because in the case that we cede any control back to JavaScript, we end up with an `IO.never` basically.

Reply:
> in develop mode didn't seem to be triggering every so often in Firefox

hmm, I wasn't able to reproduce this with Firefox 115.0.2. Were you using the latest Cats Effect?

---

As an aside, `IOApp` isn't particularly meaningful for web apps. See:
- https://github.com/typelevel/cats-effect/pull/3355
- https://armanbilge.github.io/calico/iowebapp.html

Reply:
`3.5.0` with Firefox `111.0.1` originally. I've got a bit of a weird setup at the moment, so while I'm actively developing against a local fork of `fs2-dom` I can keep an eye out to see if it happens again while using the `cats-effect` runtime?

Reply:
Yes, I would appreciate that! I'm surprised it happened at all.